src/SystemContainer.hpp

```diff
@@ -0,0 +1,162 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef GZ_SIM_SYSTEM_CONTAINER_HH_
+#define GZ_SIM_SYSTEM_CONTAINER_HH_
+
+#include <vector>
+#include <cstddef>
+#include <functional>
+
+namespace gz
+{
+  namespace sim
+  {
+    //////////////////////////////////////////////////
+    /// This container implements a simple masked vector.
+    /// Using a masked vector for systems ensures that
+    /// when a system is deleted, its memory location stays
+    /// the same. This is important because there are a number
+    /// of places where we refer to systems by raw pointer
+    /// values.
+    ///
+    /// Operation times:
+    /// push - O(1)
+    /// removeIf - O(n)
+    /// iterate - O(n) where n is the max number of systems
+    ///  -> TODO(arjoc): It should be possible to use a cache
+    ///  to make each iterator increment have O(1) time.
+    ///  For now worst case iterator increment is O(n).
+    ///  That being said for most cases the increment
+    ///  should be O(1). Eitherways, run does not iterate
+    ///  through systems but rather ISystem* interfaces.
+    ///  The only two places we iterate over this is during
+    ///  reset and TotalByEntity in the tests.
+    template<typename T>
+    class SystemContainer
+    {
+      //////////////////////////////////////////
+      /// Push an item onto the container
+      public: void Push(T internal)
+      {
+        if (freeSpots.size() == 0)
+        {
+          this->systems.push_back(internal);
+          this->occupied.push_back(true);
+          return;
+        }
+
+        auto freeIdx = freeSpots.back();
+        freeSpots.pop_back();
+        this->systems[freeIdx] = internal;
+        this->occupied[freeIdx] = true;
+      }
+
+      //////////////////////////////////////////
+      /// Clear the container
+      public: void Clear()
+      {
+        this->systems.clear();
+        this->freeSpots.clear();
+        this->occupied.clear();
+      }
+
+      //////////////////////////////////////////
+      /// Number of active systems
+      public: std::size_t Size() const
+      {
+        return this->systems.size() - this->freeSpots.size();
+      }
+
+      //////////////////////////////////////////
+      /// Remove an item if a condition is met.
+      public: void RemoveIf(std::function<bool(const T&)> fn)
+      {
+        for (std::size_t i = 0; i < systems.size(); i++)
+        {
+          if (occupied[i] && fn(systems[i]))
+          {
+            occupied[i] = false;
+            freeSpots.push_back(i);
+          }
+        }
+      }
+
+      private: std::vector<T> systems;
+
+      private: std::vector<bool> occupied;
+
+      private: std::vector<std::size_t> freeSpots;
+
+      //////////////////////////////////////////
+      class iterator : public std::iterator<
```

Comment:
I just noticed clang is giving deprecation warnings about using `std::iterator`, which is [deprecated as of c++17](https://en.cppreference.com/w/cpp/iterator/iterator)

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=gz_sim-ci-pr_any-homebrew-amd64&build=429)](https://build.osrfoundation.org/job/gz_sim-ci-pr_any-homebrew-amd64/429/) https://build.osrfoundation.org/job/gz_sim-ci-pr_any-homebrew-amd64/429/clang/

Reply:
[5001971](https://github.com/gazebosim/gz-sim/pull/2232/commits/500197164f398279d65028af9a9549bfda48c6a8)

Reply:
[5001971](https://github.com/gazebosim/gz-sim/pull/2232/commits/500197164f398279d65028af9a9549bfda48c6a8)